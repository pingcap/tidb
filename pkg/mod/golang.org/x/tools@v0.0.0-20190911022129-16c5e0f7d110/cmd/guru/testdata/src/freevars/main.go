package main

// Tests of 'freevars' query.
// See go.tools/guru/guru_test.go for explanation.
// See freevars.golden for expected query results.

// TODO(adonovan): it's hard to test this query in a single line of gofmt'd code.

type T struct {
	a, b int
}

type S struct {
	x int
	t T
}

func f(int) {}

func main() {
	type C int
	x := 1
	const exp = 6
	if y := 2; x+y+int(C(3)) != exp { // @freevars fv1 "if.*{"
		panic("expected 6")
	}

	var s S

	for x, y := range "foo" {
		println(s.x + s.t.a + s.t.b + x + int(y)) // @freevars fv2 "print.*y."
	}

	f(x) // @freevars fv3 "f.x."

loop: // @freevars fv-def-label "loop:"
	for {
		break loop // @freevars fv-ref-label "break loop"
	}
}
