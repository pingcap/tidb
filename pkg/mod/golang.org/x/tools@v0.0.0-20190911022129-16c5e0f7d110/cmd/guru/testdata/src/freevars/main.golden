-------- @freevars fv1 --------
Free identifiers:
type C
const exp int
var x int

-------- @freevars fv2 --------
Free identifiers:
var s.t.a int
var s.t.b int
var s.x int
var x int
var y rune

-------- @freevars fv3 --------
Free identifiers:
var x int

-------- @freevars fv-def-label --------
No free identifiers.

-------- @freevars fv-ref-label --------
Free identifiers:
label loop

